### DISCUSSION TITLE
[General] gasLimit in estimateGas with paymaster

### DISCUSSION THREAD

**Discussion by zonghai-li** [zkSync-Community-Hub/zksync-developers#397] [General] [answered]
### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement.

In front end, trying to estimate gas of a transaction paid by paymasters using "zksync-ethers" SDK. The snippet:

```
...
const paymasterParams = utils.getPaymasterParams(...);  // ERC20 ApprovalBased

provider.gasEstimate({
      customData: {
        gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
        paymasterParams,
      },
        value: 10_000_000_000_000n,
        from: '0x578598794419...',
        to: '0xdba5ab2e019657...,
    });
```

### What is the specific issue or error you're encountering?

The `Provider.estimateGas()` always requires the caller to hold minimum payment tokens of equivalent value of 0.008001ETH, otherwise the estimate call fails with insufficient token error. 

Ruled out the case where the original transaction(without paymaster) may cause the issue. 

Tried with different well known paymaster implementation on mainnet. The result is not depending on the paymaster contract. though the error message may differ. 

### Can you share the error messages or logs you're receiving, if any?

```
...
invocation: null,
  revert: null,
  shortMessage: 'missing revert data',
  info: {
    error: {
      code: 3,
      message: 'failed paymaster validation. error message: Failed to transfer: 6075619171527689919, 8001000000000000',
      data: '0x'
    },
```

### Have you made any recent changes to the contract before encountering this issue?

na

### Are there any external libraries or contracts that your contract interacts with?

tested with this paymaster contract, for this contract prints out the requiredToken and requiredETH on payment token transfer error.
https://explorer.zksync.io/address/0x4ae2Ba9A5C653038C6d2F5D9F80B28011A454597#contract

the validateAndPayForPaymasterTransaction function is copied here:

```
function validateAndPayForPaymasterTransaction(
        bytes32,
        bytes32,
        Transaction calldata _transaction
    )
        external
        payable
        onlyBootloader
        returns (bytes4 magic, bytes memory context)
    {
        // By default we consider the transaction as accepted.
        magic = PAYMASTER_VALIDATION_SUCCESS_MAGIC;
        require(
            _transaction.paymasterInput.length >= 4,
            "The standard paymaster input must be at least 4 bytes long"
        );
        bytes4 paymasterInputSelector = bytes4(
            _transaction.paymasterInput[0:4]
        );
        if (paymasterInputSelector == IPaymasterFlow.approvalBased.selector) {
            // While the transaction data consists of address, uint256 and bytes data,
            // the data is not needed for this paymaster
            (address token, ,) = abi.decode(_transaction.paymasterInput[4:],(address, uint256, bytes));

            AllowedTokens memory _tokenSupport = enabledTokens[token];
            // Verify if token is enabled
            require(_tokenSupport.enabled, "Invalid token");

            // We verify that the user has provided enough allowance
            address userAddress = address(uint160(_transaction.from));

            address thisAddress = address(this);

            uint256 providedAllowance = IERC20(token).allowance(
                userAddress,
                thisAddress
            );

            // Note, that while the minimal amount of ETH needed is tx.gasPrice * tx.gasLimit,
            // neither paymaster nor account are allowed to access this context variable.
            uint256 requiredETH = _transaction.gasLimit * _transaction.maxFeePerGas;

            uint256 requiredTokens;

            if(_tokenSupport.lpPair != address(0))
                requiredTokens = amountNeeded(requiredETH, _tokenSupport.lpPair);
            else
                requiredTokens = _tokenSupport.fixedPrice * requiredETH / 1e18;

            require(providedAllowance >= requiredTokens, "Min allowance too low");

            try
                IERC20(token).transferFrom(userAddress, thisAddress, requiredTokens)
            {} catch (bytes memory) {
                string memory _err = string(abi.encodePacked("Failed to transfer: ", Strings.toString(requiredTokens), ", ", Strings.toString(requiredETH)));
                revert(_err);  
            }

            // The bootloader never returns any data, so it can safely be ignored here.
            (bool success, ) = payable(BOOTLOADER_FORMAL_ADDRESS).call{
                value: requiredETH
            }("");
            require(
                success,
                "Failed to transfer tx fee to the bootloader. Paymaster balance might not be enough."
            );

            // set gas price & fee in context to refund gas amounts
            context = abi.encode(requiredETH, requiredTokens);

        } else {
            revert("Unsupported paymaster flow");
        }
    }
```

### Can you provide the relevant portions of your contract code where the issue is occurring?

na

### Have you tried to isolate the problem, and if so, what were the results?

na

### What steps have you already taken to try to resolve the issue?

in calculating the requredETH in paymaster contract, 

`
 uint256 requiredETH = _transaction.gasLimit * _transaction.maxFeePerGas;
`

it depends on the gasLimit and gasPrice of the transaction. If we provide an initial value to gasLimit or maxFeePerGas in the estimateGas call, the provided values don't take effect. It seems these values are set constant during the estimation workflow, which I guess the value for gasLimit is set to 80010000.(remember the boot loader address is 0x8001)

The discussion: is it the case that gasLimit is set to a default value during the estimation pass? To make it more tolerant with user's holding asset value, how would front-end program provide an gasLimit value in the estimateGas call?

**Comment by uF4No** [comment, marked as answer]
Hi @zonghai-li 

The `estimateGas` method does not require the sender to hold any assets. Here is part of a script to estimate gas for a tx sent via paymaster on an empty account which just returns the estimated gas as expected:

```js
const emptyWallet = Wallet.createRandom();
const userWallet = new Wallet(emptyWallet.privateKey, provider);

const gasPrice = await provider.getGasPrice();
const paymasterParams = utils.getPaymasterParams(paymasterAddress, {
  type: "General",
  innerInput: new Uint8Array(),
});

await greeter.connect(userWallet);

const gasLimit = await greeter.setGreeting.estimateGas("Hola, mundo!", {maxPriorityFeePerGas: BigInt(0),
   maxFeePerGas: gasPrice,
  customData: {
     gasPerPubdata: utils.DEFAULT_GAS_PER_PUBDATA_LIMIT,
     paymasterParams,
}})
console.log('gasLimit :>> ', gasLimit);
```

Output

```sh
gasLimit :>>  1497734n
```

**Reply by zonghai-li** [reply]
Thanks for looking into it!

This takes place when it comes to the ApprovalBased workflow, where an amount of payment tokens need to be transferred from the user based on the required ETH gas fee. If we look into the code in paymaster contract 
```
 // Note, that while the minimal amount of ETH needed is tx.gasPrice * tx.gasLimit,
            // neither paymaster nor account are allowed to access this context variable.
            uint256 requiredETH = _transaction.gasLimit * _transaction.maxFeePerGas;

 ```
we see that the `requiredETH` is calculated from `transaction.gasLimit` and `maxFeePerGas`.  It is interesting that `transaction.gasLimit` seems to always take constant value during in the gas estimating pass, thus resulting in a constant requiredETH, which is 0.008001 Ether from my observation. Since the required gas needs to be charged from the user in the chosen payment token, when the amount held by user is less than the required value,  it ends up the estimateGas() fails with revert error. 

I tested this paymaster on mainnet
https://explorer.zksync.io/address/0x4ae2Ba9A5C653038C6d2F5D9F80B28011A454597#contract
with the $HOLD token. 

I wonder if people can also confirm this scenario. If so, this minimum requirement does really matters to the end users? 0.008001 Ether corresponds to $30 as of today. While an actual transaction fee is probably less than $1 on zksync. It is possible(or even makes sense) to specify a gasLimit or maxFeePerGas in the transaction request body of estimateGas() to make it far closer to the actual required gas fee. 


--

**Reply by uF4No** [reply]
Hi

Not sure if the provided paymaster has additional logic but the only account that has to have ETH is the paymaster itself as it's the one that will pay the gas fee in ETH to the bootloader.

We have a kinda [similar tutorial here](https://docs.zksync.io/build/tutorials/smart-contract-development/paymasters/erc20fixed.html). It's an `approvalBased` paymaster that requires a unit of an ERC20 token in exchange of the gas fees. You can interact with it from an account with 0ETH and it'll work ok.

Hope that helps
